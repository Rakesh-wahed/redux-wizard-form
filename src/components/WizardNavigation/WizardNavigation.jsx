// @flow
import * as React from 'react';
import {
  Wrapper,
  Block,
  ProgressBarWrapper,
  ProgressBar,
  NumberStep
} from './styles';

type Props = {
  currentStep: number,
  stepsSize: number,
  stepsNames: Array<string>,
  children?: Function,
  onGoToStep: Function
};

export const WizardNavigationComponent = ({
  currentStep,
  stepsSize,
  stepsNames,
  children,
  onGoToStep
}: Props): React.Node =>
  typeof children !== 'function' ? (
    <Wrapper>
      <ProgressBarWrapper stepsSize={stepsSize}>
        <ProgressBar currentStep={currentStep} stepsSize={stepsSize} />
      </ProgressBarWrapper>
      {stepsNames.map((stepName, i) => (
        <Block key={i} onClick={() => onGoToStep(currentStep, i)}>
          <NumberStep isActive={currentStep >= i}>
            <span>{i + 1}</span>
          </NumberStep>
          <span className="step-name">{stepName}</span>
        </Block>
      ))}
    </Wrapper>
  ) : (
    children(currentStep, stepsSize, stepsNames)
  );

WizardNavigationComponent.defaultProps = {
  children: null
};
