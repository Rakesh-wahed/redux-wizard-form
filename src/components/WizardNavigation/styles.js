import styled from 'styled-components';

const barColor = '#00bcd4';

export const Wrapper = styled.div`
  align-items: center;
  display: flex;
  justify-content: space-around;
  margin-bottom: 20px;
  position: relative;
  width: 100%;
`;

export const ProgressBarWrapper = styled.div`
  left: ${props => 100 / props.stepsSize / 2}%;
  height: 3px;
  top: 13px;
  position: absolute;
  width: ${props => 100 - 100 / props.stepsSize}%;
`;

export const Block = styled.div`
  align-items: center;
  background: transparent;
  cursor: pointer;
  display: flex;
  flex-direction: column;
  justify-content: center;
  text-align: center;
  width: 100%;

  span {
    font-size: 1rem;
  }

  .step-name {
    overflow: hidden;
    text-overflow: ellipsis;
    white-space: nowrap;
    width: 90%;
  }
`;

export const ProgressBar = styled.div`
  background: ${barColor};
  border-radius: 10px;
  height: 100%;
  transition: width 600ms cubic-bezier(0, 0.89, 0.69, 0.98);
  width: ${props => props.currentStep / (props.stepsSize - 1) * 100}%;
`;

export const NumberStep = styled.div`
  background-color: ${props => (props.isActive ? barColor : '#9E9E9E')};
  border-radius: 50%;
  color: white;
  font-weight: 800;
  height: 20px;
  overflow: hidden;
  position: relative;
  padding: 5px;
  margin-bottom: 10px;
  text-align: center;
  transition: background-color 300ms cubic-bezier(0, 0.89, 0.69, 0.98);
  width: 20px;
`;
