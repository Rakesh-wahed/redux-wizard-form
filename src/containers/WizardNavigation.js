import { connect } from 'react-redux';
import { getCurrentStep, getStepsSize } from '../reducer';
import { WizardNavigationComponent } from '../components/WizardNavigation/WizardNavigation.jsx';

export const WizardNavigation = connect(state => ({
  currentStep: getCurrentStep(state),
  stepsSize: getStepsSize(state)
}))(WizardNavigationComponent);
