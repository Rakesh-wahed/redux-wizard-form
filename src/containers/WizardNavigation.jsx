// @flow
import { connect } from 'react-redux';
import { getCurrentStep, getStepsSize, getStepsNames } from '../reducer';
import { WizardNavigationComponent } from '../components/WizardNavigation/WizardNavigation';
import { goToStep } from '../actions';

export const WizardNavigation = connect(
  (state: Object): Object => ({
    currentStep: getCurrentStep(state),
    stepsSize: getStepsSize(state),
    stepsNames: getStepsNames(state)
  }),
  (dispatch: Dispatch): Object => ({
    onGoToStep: (currentStep: number, step: number): void => {
      if (step < currentStep) {
        dispatch(goToStep(step));
      }
    }
  })
)(WizardNavigationComponent);
