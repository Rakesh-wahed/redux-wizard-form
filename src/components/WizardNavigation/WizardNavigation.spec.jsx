import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { WizardNavigationComponent } from './WizardNavigation';

Enzyme.configure({ adapter: new Adapter() });

describe('WizardNavigation component', () => {
  test('snapshot', () => {
    const props = { currentStep: 0, stepsSize: 1, stepsNames: [], onGoToStep: () => {} };
    const Component = shallow(<WizardNavigationComponent {...props} />);
    expect(Component).toMatchSnapshot();
  });
});
