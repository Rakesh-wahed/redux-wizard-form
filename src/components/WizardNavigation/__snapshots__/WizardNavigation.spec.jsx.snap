// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WizardNavigation component snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <WizardNavigationComponent
    currentStep={0}
    onGoToStep={[Function]}
    stepsNames={Array []}
    stepsSize={1}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <styled.div
          stepsSize={1}
        >
          <styled.div
            currentStep={0}
            stepsSize={1}
          />
        </styled.div>,
        Array [],
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": <styled.div
            currentStep={0}
            stepsSize={1}
          />,
          "stepsSize": 1,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "currentStep": 0,
            "stepsSize": 1,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <styled.div
            stepsSize={1}
          >
            <styled.div
              currentStep={0}
              stepsSize={1}
            />
          </styled.div>,
          Array [],
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": <styled.div
              currentStep={0}
              stepsSize={1}
            />,
            "stepsSize": 1,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "currentStep": 0,
              "stepsSize": 1,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
